ichnaea/api/locate/tests/base.py

```diff
@@ -701,96 +764,118 @@ def test_cellarea_with_different_fallback(self, app, session, stats):
+                "fallback_allowed:false",
+                "accuracy:low",
+                "status:hit",
+                "source:internal",
+            ],
         )
 
-    def test_wifi_not_found(self, app, data_queues, stats):
+    def test_wifi_not_found(self, app, data_queues, metricsmock):
         wifis = WifiShardFactory.build_batch(2)
 
         query = self.model_query(wifis=wifis)
 
         res = self._call(app, body=query, status=self.not_found.code)
         self.check_response(data_queues, res, "not_found")
-        stats.check(
-            counter=[
-                (
-                    "request",
-                    [
-                        self.metric_path,
-                        "method:post",
-                        "status:%s" % self.not_found.code,
-                    ],
-                ),
-                (self.metric_type + ".request", [self.metric_path, "key:test"]),
-                (
-                    self.metric_type + ".query",
-                    [
-                        "key:test",
-                        "region:none",
-                        "geoip:false",
-                        "blue:none",
-                        "cell:none",
-                        "wifi:many",
-                    ],
-                ),
-                (
-                    self.metric_type + ".result",
-                    "fallback_allowed:false",
-                    ["key:test", "region:none", "accuracy:high", "status:miss"],
-                ),
-                (
-                    self.metric_type + ".source",
-                    [
-                        "key:test",
-                        "region:none",
-                        "source:internal",
-                        "accuracy:high",
-                        "status:miss",
-                    ],
-                ),
+        assert metricsmock.has_record(
+            "incr",
+            "request",
+            value=1,
+            tags=[self.metric_path, "method:post", "status:%s" % self.not_found.code],
+        )
+        assert metricsmock.has_record(
+            "incr",
+            self.metric_type + ".request",
+            value=1,
+            tags=[self.metric_path, "key:test"],
+        )
+        assert metricsmock.has_record(
+            "incr",
+            self.metric_type + ".query",
+            value=1,
+            tags=[
+                "key:test",
+                "region:none",
+                "geoip:false",
+                "blue:none",
+                "cell:none",
+                "wifi:many",
             ],
-            timer=[("request", [self.metric_path, "method:post"])],
+        )
+        metricsmock.print_records()
+        assert metricsmock.has_record(
+            "incr",
+            self.metric_type + ".result",
+            value=1,
+            tags=[
+                "key:test",
+                "region:none",
+                "accuracy:high",
+                "fallback_allowed:false",
```

Comment:
This one is curious. The original check argument is this:

```
(
    self.metric_type + ".result",
    "fallback_allowed:false",
    ["key:test", "region:none", "accuracy:high", "status:miss"],
),
```

The check code is here:

https://github.com/mozilla/ichnaea/blob/ba487a97bca4b6343ae0365267ea8cdd6de3f553/ichnaea/log.py#L275-L309

That's the three arguments possibility for check. So `self.metric_type + ".result"` is the name, `"fallback_allowed:false"` is the match, and the value is the list of tags. The code then checks to see if the value is a list and if so, that becomes the tags and the value becomes `None`.

The match gets ignored because it's not an int. Seems like this is a bug. There are a couple of other instances of this--but not many in the grand scheme of things.

Anyhow, `.check()` checks to see if the listed tags are a subset of the tags of the metrics message. In this case, it's a subset, so `.check()` works fine.

However, Markus checks to see if the expected list of tags is exactly the same as the actual list of tags and it's not because `"fallback_allowed:false"` isn't in the expected list. So I added that and now the test passes.

Why write all this out? Because it's not straight-forward and it was curious and while I think I got it right, maybe I didn't.

Reply:
I think this was a mistake as well. It was added in 2015 in https://github.com/mozilla/ichnaea/commit/add1bacceff77b0cd4706590a42bea6fb5dddfef, and there are many similar tests that got updated to add ``fallback_allowed:false``. Most correctly add to the tag list, but there are a sequence of them in this file where it got added in the wrong place.  It is a shame that there were no assertions for a missing tag in the list, or that ``match`` was a non integer. Glad to see this fixed!